pkg/kwokctl/cmd/stop/cluster/cluster.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spf13/cobra"
+
+	"sigs.k8s.io/kwok/pkg/kwokctl/runtime"
+	"sigs.k8s.io/kwok/pkg/kwokctl/utils"
+	"sigs.k8s.io/kwok/pkg/kwokctl/vars"
+	"sigs.k8s.io/kwok/pkg/log"
+)
+
+type flagpole struct {
+	Name string
+}
+
+// NewCommand returns a new cobra.Command for stop cluster
+func NewCommand() *cobra.Command {
+	flags := &flagpole{}
+	cmd := &cobra.Command{
+		Args:  cobra.NoArgs,
+		Use:   "cluster",
+		Short: "Stop a cluster",
+		Long:  "Stop a cluster",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			flags.Name = vars.DefaultCluster
+			return runE(cmd.Context(), flags)
+		},
+	}
+	return cmd
+}
+
+func runE(ctx context.Context, flags *flagpole) error {
+	name := fmt.Sprintf("%s-%s", vars.ProjectName, flags.Name)
+	workdir := utils.PathJoin(vars.ClustersDir, flags.Name)
+
+	logger := log.FromContext(ctx)
+	logger = logger.With("cluster", flags.Name)
+	ctx = log.NewContext(ctx, logger)
+
+	rt, err := runtime.DefaultRegistry.Load(name, workdir)
+	if err != nil {
+		return err
+	}
+
+	etcdDataTmp := utils.PathJoin(workdir, vars.DefaultSnapshotFilename)
+	err = rt.SnapshotSave(ctx, etcdDataTmp)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("Stopping cluster")
+	err = rt.Down(ctx)
```

Comment:
https://github.com/kubernetes-sigs/kwok/blob/9266eef627d3c803c4af0a78710a6a1b09bf2e96/pkg/kwokctl/runtime/kind/cluster.go#L242-L262

I'm not quite sure if it will work properly if it's a Kind runtime

Reply:
I tested it locally, and it works
```sh
➜  ~ kubectl config use-context kwok-kwok
Switched to context "kwok-kwok".
➜  ~ k get po
NAME                        READY   STATUS    RESTARTS   AGE
fake-pod-759cdcff4c-cft9s   1/1     Running   0          11h
fake-pod-759cdcff4c-l947m   1/1     Running   0          11h
fake-pod-759cdcff4c-mbxn9   1/1     Running   0          11h
fake-pod-759cdcff4c-mht8f   1/1     Running   0          11h
fake-pod-759cdcff4c-mvb44   1/1     Running   0          11h
fake-pod-759cdcff4c-pnw74   1/1     Running   0          11h
fake-pod-759cdcff4c-r42bv   1/1     Running   0          11h
fake-pod-759cdcff4c-xmbch   1/1     Running   0          11h
fake-pod-759cdcff4c-z4s6f   1/1     Running   0          11h
fake-pod-759cdcff4c-zlvgb   1/1     Running   0          11h
➜  ~ kwokctl stop cluster --name kwok
Stopping cluster                                                    cluster=kwok
Deleting cluster "kwok-kwok" ...
Cluster stopped                                                     cluster=kwok
➜  ~ kwokctl start cluster --name kwok
Starting cluster                                                    cluster=kwok
Creating cluster "kwok-kwok" ...
 ✓ Ensuring node image (kindest/node:v1.21.1) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kwok-kwok"
You can now use your cluster with:

kubectl cluster-info --context kind-kwok-kwok

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
Cluster started                                                     cluster=kwok
➜  ~ k get po
NAME                        READY   STATUS    RESTARTS   AGE
fake-pod-759cdcff4c-cft9s   1/1     Running   0          11h
fake-pod-759cdcff4c-l947m   1/1     Running   0          11h
fake-pod-759cdcff4c-mbxn9   1/1     Running   0          11h
fake-pod-759cdcff4c-mht8f   1/1     Running   0          11h
fake-pod-759cdcff4c-mvb44   1/1     Running   0          11h
fake-pod-759cdcff4c-pnw74   1/1     Running   0          11h
fake-pod-759cdcff4c-r42bv   1/1     Running   0          11h
fake-pod-759cdcff4c-xmbch   1/1     Running   0          11h
fake-pod-759cdcff4c-z4s6f   1/1     Running   0          11h
fake-pod-759cdcff4c-zlvgb   1/1     Running   0          11h
```